apps/web/src/components/tx/SignOrExecuteForm/SignOrExecuteForm.tsx

```diff
@@ -124,6 +127,55 @@ export const SignOrExecuteForm = ({
     [onFormSubmit],
   )
 
+  const onNoteSubmit = useCallback(
+    (note: string) => {
+      const originalOrigin = props.origin ? JSON.parse(props.origin) : { url: location.origin }
+      setCustomOrigin(
+        JSON.stringify({
+          ...originalOrigin,
+          name: JSON.stringify({ note }),
+        }),
+      )
+    },
+    [setCustomOrigin, props.origin],
+  )
+
+  const getForm = () => {
+    const commonProps = {
+      ...props,
+      safeTx,
+      isCreation,
+      origin: customOrigin,
+      onSubmit: onFormSubmit,
+    }
+    if (isCounterfactualSafe && !isProposing) {
+      return <CounterfactualForm {...commonProps} onlyExecute />
+    }
+
+    if (!isCounterfactualSafe && willExecute && !isProposing) {
+      return <ExecuteForm {...commonProps} />
+    }
+
+    if (!isCounterfactualSafe && willExecuteThroughRole) {
+      return (
+        <ExecuteThroughRoleForm
+          {...commonProps}
+          role={(allowingRole || mostLikelyRole)!}
+          safeTxError={safeTxError}
+          onSubmit={onRoleExecutionSubmit}
+        />
+      )
+    }
+
+    if (!isCounterfactualSafe && !willExecute && !willExecuteThroughRole && !isProposing) {
+      return <SignForm {...commonProps} isBatchable={isBatchable} />
+    }
+
```

Comment:
One suggestion to improve this bunch of "ifs" later would be adopt an buiilder pattern here. Where you could have an mapper where the key would be the builder name and the value the component you wanna build. And to render the builder you could use a director to receive the builder name and map it according to the value you find inside the map... Just an idea. not a blocker though.

For reference https://refactoring.guru/design-patterns/builder